neurolib/models/multimodel/builder/model_input.py

```diff
@@ -0,0 +1,310 @@
+"""
+Handles input to neural models. Constructs both noisy and stimulation-like input
+and supports both CubicHermiteSplines for jitcdde backend and np.array for numba
+backend.
+"""
+
+import numba
+import numpy as np
+from chspy import CubicHermiteSpline
+from scipy.signal import square
+
+
+class ModelInput:
+    """
+    Generates input to neural model.
+    """
+
+    def __init__(self, duration, dt, independent_realisations=1, seed=None):
+        """
+        :param duration: duration of the input, in miliseconds
+        :type duration: float
+        :param dt: some reasonable "speed" of input, in miliseconds
+        :type dt: float
+        :param independent_realisations: how many independent realisation of
+            the input we want - for constant inputs the array is just copied,
+            for noise this means independent realisation
+        :type independent_realisations: int
+        :param seed: optional seed for noise generator
+        :type seed: int|None
+        """
+        self.times = np.arange(dt, duration + dt, dt)
+        self.dt = dt
+        self.num_iid = independent_realisations
+        # seed the generator
+        np.random.seed(seed)
+
+    def generate_input(self):
+        """
+        Function to generate input.
+        """
+        raise NotImplementedError
+
+    def as_array(self):
+        """
+        Return input as numpy array.
+        """
+        return self.generate_input()
+
+    def as_cubic_splines(self):
+        """
+        Return as cubic Hermite splines.
+        """
+        return CubicHermiteSpline.from_data(self.times, self.generate_input())
+
+
+class StimulusInput(ModelInput):
+    """
+    Generates stimulus input with optional start and end times.
+    """
+
+    def __init__(
+        self, duration, dt, stim_start=None, stim_end=None, independent_realisations=1, seed=None,
+    ):
+        """
+        :param stim_start: start of the stimulus, in miliseconds
+        :type stim_start: float
+        :param stim_end: end of the stimulus, in miliseconds
+        :type stim_end: float
+        """
+        self.stim_start = stim_start or 0.0
+        self.stim_end = stim_end or duration
+        assert self.stim_start < duration
+        assert self.stim_end <= duration
+        super().__init__(
+            duration=duration, dt=dt, independent_realisations=independent_realisations, seed=seed,
+        )
+
+    def as_array(self):
+        """
+        Return input as numpy array after checking stimulus bounds.
+        """
+        stim_input = self.generate_input()
+        stim_input[self.times < self.stim_start] = 0.0
+        stim_input[self.times > self.stim_end] = 0.0
+        return stim_input
+
+    def as_cubic_splines(self):
```

Comment:
this method is for `jitcdde` backend with adaptive `dt` - with adaptive `dt` and delays and you need to provide system input in a way in which the integrator can interpolate the value for an input in arbitrary time - this uses cubic Hermite splines